scripts/schema/finalizer.py

```diff
@@ -0,0 +1,171 @@
+import copy
+
+from schema import visitor
+
+# This script takes the fleshed out deeply nested fields dictionary as emitted by
+# cleaner.py, and performs field reuse in two phases.
+#
+# Phase 1 performs field reuse across field sets. E.g. `group` fields should also be under `user`.
+# This type of reuse is then carried around if the receiving field set is also reused.
+# In other words, user.group.* will be in other places where user is nested:
+# source.user.* will contain source.user.group.*
+
+# Phase 2 performs field reuse where field sets are reused within themselves, with a different name.
+# Examples are nesting `process` within itself, as `process.parent.*`,
+# or nesting `user` within itself at `user.target.*`.
+# This second kind of nesting is not carried around everywhere else the receiving field set is reused.
+# So `user.target.*` is *not* carried over to `source.user.target*` when we reuse `user` under `source`.
+
+
+def finalize(fields):
+    '''Intended entrypoint of the finalizer.'''
+    perform_reuse(fields)
+    calculate_final_values(fields)
+
+
+def order_reuses(fields):
```

Comment:
Yes I totally agree this is a great job for computers ;-)

At this time, `user` is the single field set that has both fields nested inside it (`group`), and is being reused elsewhere.

This is the reason why I didn't go there yet. That's a lot of code to handle exactly one occurrence.

Note that this reuse "order" has no impact at all on any other field set reuse.

Do you think it's a blocker for this PR?

Reply:
To me it seems like a blocker since it would be a breaking change for existing custom schemas. The custom schemas we're using for the endpoint alerts also have multiple layers of reused fields.

If you'd like I can come up with a proposed implementation and make a PR to your branch with it.

Reply:
As discussed elsewhere, this will be done as a separate PR.